### PR TITLE
Obsolete LeakTrackingObjectPool and LeakTrackingObjectPoolProvider

### PR DESCRIPTION
**Pull request by adityamandaleeka**
`LeakTrackingObjectPool` doesn't work on release builds, and nothing else depends on it, so we've chosen to mark it (and `LeakTrackingObjectPoolProvider`) obsolete.

Address https://github.com/dotnet/aspnetcore/issues/42524

### CODE REVIEW

File: src/ObjectPool/src/LeakTrackingObjectPool.cs [hunk 1]
```diff
@@ -16,6 +16,7 @@ namespace Microsoft.Extensions.ObjectPool;
 /// </para>
 /// </summary>
 /// <typeparam name="T">The type of object which is being pooled.</typeparam>
+[Obsolete("LeakTrackingObjectPool<T> was only intended for use in diagnostic builds and may be removed in a future release.")]
```
Comment (Jehoel) [inline src/ObjectPool/src/LeakTrackingObjectPool.cs:19]: The text "was only intended for use in diagnostic builds" can be interpreted as referring to the user's diag builds, not Microsoft's builds. 

I'd phrase it like this, perhaps:

> "LeakTrackingObjectPool<T> is only functional in diagnostic builds of the .NET CLR and EE. It does not function in GA/release builds of .NET used by the general public. This type may be removed in a future release."

Comment (adityamandaleeka) [reply src/ObjectPool/src/LeakTrackingObjectPool.cs:19]: Thanks @Jehoel. It doesn't actually need a special CLR build to work, but I understand your feedback. How about:

> `LeakTrackingObjectPool<T>` was only intended for use in diagnostic builds of ASP.NET Core. It does not function in any publicly shipped ASP.NET Core versions and may be removed in a future release.

Comment (Jehoel) [reply src/ObjectPool/src/LeakTrackingObjectPool.cs:19]: 👍

File: src/ObjectPool/src/LeakTrackingObjectPool.cs [hunk 2]
```diff
@@ -16,6 +16,8 @@ namespace Microsoft.Extensions.ObjectPool;
 /// </para>
 /// </summary>
 /// <typeparam name="T">The type of object which is being pooled.</typeparam>
+[Obsolete("LeakTrackingObjectPool<T> was only intended for use in diagnostic builds of .NET. It does not "+
+    "function in any publicly shipped .NET versions and may be removed in a future release.")]
```
Comment (halter73) [inline src/ObjectPool/src/LeakTrackingObjectPool.cs:20]: Nit:

```suggestion
[Obsolete("LeakTrackingObjectPool<T> was only intended for internal use in diagnostic builds of .NET. It does not "+
    "function in any publicly shipped .NET versions and may be removed in a future release.")]
```

File: src/ObjectPool/src/LeakTrackingObjectPoolProvider.cs [hunk 3]
```diff
@@ -9,6 +9,8 @@ namespace Microsoft.Extensions.ObjectPool;
 /// An <see cref="ObjectPoolProvider"/> that produces instances of
 /// <see cref="LeakTrackingObjectPool{T}"/>.
 /// </summary>
+[Obsolete("LeakTrackingObjectPoolProvider was only intended for use in diagnostic builds of .NET. It does not "+
```
Comment (halter73) [inline src/ObjectPool/src/LeakTrackingObjectPoolProvider.cs:12]:
```suggestion
[Obsolete("LeakTrackingObjectPoolProvider was only intended for internal use in diagnostic builds of .NET. It does not "+
```

**Review by halter73 — APPROVED** [review]

File: src/ObjectPool/src/LeakTrackingObjectPool.cs [hunk 4]
```diff
@@ -16,6 +16,8 @@ namespace Microsoft.Extensions.ObjectPool;
 /// </para>
 /// </summary>
 /// <typeparam name="T">The type of object which is being pooled.</typeparam>
+[Obsolete("LeakTrackingObjectPool<T> was only intended for internal use in diagnostic builds of .NET. It does not " +
+    "function in any publicly shipped .NET versions and may be removed in a future release.")]
```
Comment (halter73) [inline src/ObjectPool/src/LeakTrackingObjectPool.cs:20]:
```suggestion
[Obsolete("LeakTrackingObjectPool<T> was only intended for internal use in diagnostic builds of .NET. It never " +
    "functioned in any publicly shipped .NET versions and may be removed in a future release.")]
```

File: src/ObjectPool/src/LeakTrackingObjectPoolProvider.cs [hunk 5]
```diff
@@ -9,6 +9,8 @@ namespace Microsoft.Extensions.ObjectPool;
 /// An <see cref="ObjectPoolProvider"/> that produces instances of
 /// <see cref="LeakTrackingObjectPool{T}"/>.
 /// </summary>
+[Obsolete("LeakTrackingObjectPoolProvider was only intended for internal use in diagnostic builds of .NET. It does not " +
+    "function in any publicly shipped .NET versions and may be removed in a future release.")]
```
Comment (halter73) [inline src/ObjectPool/src/LeakTrackingObjectPoolProvider.cs:13]:
```suggestion
[Obsolete("LeakTrackingObjectPoolProvider was only intended for internal use in diagnostic builds of .NET. It never " +
    "functioned in any publicly shipped .NET versions and may be removed in a future release.")]
```

**Pull request merged**